src/ocean/util/container/map/TreeMap.d

```diff
@@ -0,0 +1,352 @@
+        added = added_ebnode is ebnode;
+
+        if (!added)
+            allocator.deallocate(node);
+
+        return cast(Node*)added_ebnode;
+    }
+
+    /***************************************************************************
+
+        `malloc` allocator for `Node`. Deallocating keeps a spare item as an
+        optimisation for the frequent situation of allocating an item, then
+        noticing it isn't actually needed and freeing it (this happens in
+        `put()` when a duplicate is detected).
+
+    ***************************************************************************/
+
+    private struct MallocAllocator
+    {
+        import core.stdc.stdlib: malloc, free, exit, EXIT_FAILURE;
+        import core.stdc.stdio: stderr, fputs;
+
+        /***********************************************************************
+
+            The spare node, set by `deallocate()` and used by `allocate()`.
+
+        ***********************************************************************/
+
+        Node* spare = null;
+
+        /***********************************************************************
+
+            Allocates a new or reuses the spare node. The returned node should
+            be deallocated by `deallocate()` when not used any more.
+
+            If `malloc` fails, prints out an error message and terminates the
+            process with `EXIT_FAILURE`.
+
+            Returns:
+                an initialised node ready to use.
+
+        ***********************************************************************/
+
+        Node* allocate ( )
+        {
+            if (auto node = this.spare)
+            {
+                this.spare = null;
+                return node;
+            }
+
+            if (auto node = cast(Node*)malloc(Node.sizeof))
+            {
+                *node = (*node).init;
+                return node;
+            }
+            else
+            {
+                const istring msg = "malloc(" ~ Node.sizeof.stringof ~
+                                   ") failed: Out of memory\n\0";
+                fputs(msg.ptr, stderr);
+                exit(EXIT_FAILURE);
+                assert(false);
+            }
+        }
+
+        /***********************************************************************
+
+            Deallocates node or stores it in the spare item.
+
+            Params:
+                node = the item to deallocate, previously obtained by
+                       `allocate()`
+
+        ***********************************************************************/
+
+        void deallocate ( Node* node )
+        {
+            if (this.spare is null)
+            {
+                *node = (*node).init;
+                this.spare = node;
+            }
+            else
+            {
+                free(node);
+            }
+        }
+    }
+}
+
+/*******************************************************************************
```

Comment:
Annoying ocean comment: a usage example and unittest would be nice! :grimacing:

Reply:
Sigh... ;-)

Reply:
I think we cut out the wrong API. As this stands, this map is just a map of `id`s, since AFAICS` `eb64_node` (to which we get the pointer from `Put`) can't hold any user defined value. We need the "other" API that supports any arbitrary data type, as long as they have `eb_node` as a first element.